samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java

```diff
@@ -59,30 +63,34 @@ public void onNext(M message, MessageCollector collector, TaskCoordinator coordi
     K key = thisPartialJoinFn.getKey(message);
     thisPartialJoinFn.getState().put(key, new PartialJoinMessage<>(message, System.currentTimeMillis()));
     PartialJoinMessage<JM> otherMessage = otherPartialJoinFn.getState().get(key);
-    if (otherMessage != null) {
+    long now = System.currentTimeMillis();
+    if (otherMessage != null && otherMessage.getReceivedAt() > now - ttlMs) {
       RM joinResult = thisPartialJoinFn.apply(message, otherMessage.getMessage());
       this.propagateResult(joinResult, collector, coordinator);
     }
   }
 
   @Override
   public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    long now = System.currentTimeMillis();
+
     KeyValueStore<K, PartialJoinMessage<M>> thisState = thisPartialJoinFn.getState();
     KeyValueIterator<K, PartialJoinMessage<M>> iterator = thisState.all();
     List<K> keysToRemove = new ArrayList<>();
-    long now = System.currentTimeMillis();
 
     while (iterator.hasNext()) {
       Entry<K, PartialJoinMessage<M>> entry = iterator.next();
-      if (entry.getValue().getTimestamp() < now - ttlMs) {
+      if (entry.getValue().getReceivedAt() < now - ttlMs) {
         keysToRemove.add(entry.getKey());
       } else {
         break;
       }
     }
 
+    iterator.close();
     thisState.deleteAll(keysToRemove);
 
+    LOGGER.debug("onTimer self time: {} ms", System.currentTimeMillis() - now);
```

Comment:
I'm guessing every second or so, since window controls the minimum granularity for onTimer across all operators (e.g. firing window triggers). Will change to info, we can turn it down later. Name should be present in the Logger class, will add operator ID.